### PR TITLE
Closes #2571 transform_range: implement transform_range()

### LINKED ISSUE #2571: Feature Request: Implement transform_range()
### Feature Idea

From suggestion by @bundfussr: 

_to help with ADQS programming, it would be convenient to implement a function which provides the same transformation arguments as `compute_scale()`:_

```
transform_range <- function(value,
                            source_range = NULL,
                            target_range = NULL,
                            flip_direction = FALSE,
                            outside_range = "NA"
)
```

_The `outside_range` argument determines how values outside the source range are handled. Options are setting to `NA`, issuing an error, or issuing a warning. (The new function could also be called by `compute_scale()`.) I would expect that it could be useful for other questionnaires as well._

### Relevant Input

See [here](https://pharmaverse.github.io/admiral/reference/compute_scale.html?q=compute_scale#null) for `compute_range()` inputs.

### Relevant Output

A transformed value

### Reproducible Example/Pseudo Code

```
transform_range(
  value = 2,
  source_range = c(1:5),
  target_range = c(0:100),
  flip_direction = FALSE,
  outside_range = NA
)
```
Should return `25`.

```
transform_range(
  value = 3,
  source_range = c(1:6),
  target_range = c(0:100),
  flip_direction = TRUE,
  outside_range = NA
)
```
Should return `60` i.e. `100 - 40`.

```
transform_range(
  value = 300,
  source_range = c(1:6),
  target_range = c(0:100),
  flip_direction = TRUE,
  outside_range = NA
)
```
Should return `NA`.

### PR DESCRIPTION
**Pull request by bundfussr**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Review the [Cheat Sheet](https://github.com/pharmaverse/admiral/blob/main/inst/cheatsheet/admiral_cheatsheet.pdf). Make any required updates to it by editing the file `inst/cheatsheet/admiral_cheatsheet.pptx` and re-upload a PDF and a PNG version of it to the same folder. (The PNG version can be created by taking a screenshot of the PDF version.)
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers). A Developer Notes section is available in `NEWS.md` for tracking developer-facing issues.  
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -6,6 +6,8 @@
 `AVALCATx` & `AVALCAxN`. (#2480)
 - New function `derive_vars_crit_flag()` for deriving criterion flag variables
 (`CRITy`, `CRITyFL`, `CRITyFLN`). (#2468)
+- New function `transform_scale()`  to transform values from a source range to a
```
Comment (bms63) [inline NEWS.md:9]: A new type of function to add to our list!?! https://pharmaverse.github.io/admiraldev/dev/articles/programming_strategy.html#function-names

Comment (bundfussr) [reply NEWS.md:9]: No, it's a computation. I.e., it is covered by
> `compute_` / `calculate_` / … | Functions that take vectors as input and return a vector

Comment (bms63) [reply NEWS.md:9]: should we add `transform_` so developers know they can use that as a verb?

Comment (bundfussr) [reply NEWS.md:9]: Then we would need to add `convert_` and `impute_` as well.

Comment (bms63) [reply NEWS.md:9]: I would be happy with that!

**Review by manciniedoardo — CHANGES_REQUESTED** [review]
Thanks @bundfussr - no concerns about the function or its roxygen - looks good to me.

Wondering if, for completeness, you could also please add a quick mention for the function (no code required) [here](https://pharmaverse.github.io/admiral/articles/questionnaires.html#scales-and-scores)? This might increase chances users come across it when doing questionnaires, which is where this originally came from anyway.

File: NEWS.md (hunk 1 above)
Comment (bundfussr) [reply NEWS.md:9]: OK, I have created an issue in admiraldev (https://github.com/pharmaverse/admiraldev/issues/472).

File: R/transform_range.R [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+#' @return The source linearly transformed to the target range
+#'
+#' @keywords com_bds_findings
+#'
+#' @family com_bds_findings
+#'
+#' @export
+#'
+#' @examples
+#' transform_range(
```
[the remark this answers is not in the document]
Comment (bundfussr) [reply R/transform_range.R:49]: Yes, example added

File: R/transform_range.R [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+#' )
+#'
+transform_range <- function(source,
+                            source_range,
+                            target_range,
+                            flip_direction = FALSE,
+                            outside_range = "NA") {
+  # Function argument checks
+  assert_numeric_vector(source)
+  assert_numeric_vector(source_range)
```
[the remark this answers is not in the document]
Comment (bundfussr) [reply R/transform_range.R]: Yes, I've enhanced `assert_numeric_vector()` in admiraldev (see https://github.com/pharmaverse/admiraldev/issues/473) and updated the calls here.

**Review by manciniedoardo — CHANGES_REQUESTED** [review]
Thanks @bundfussr - just one more comment from me

File: vignettes/questionnaires.Rmd [hunk 4]
```diff
@@ -123,7 +123,8 @@ Depending on the question, `QSORRES == "YES"` is mapped to `QSSTRESN = 0` or
 `QSSTRESN = 1`. If the `QSSTRESN` values are not ready to be used for deriving
 scores and require transformation, it is recommended that `QSSTRESN` is kept in
 the ADaM dataset for traceability, and the transformed value is stored in
-`AVAL`, since that's what will be used for the score calculation.
+`AVAL`, since that's what will be used for the score calculation. The
```
[the remark this answers is not in the document]
Comment (bundfussr) [reply vignettes/questionnaires.Rmd]: OK, I've added a "Transformed Items" section.

**Review by manciniedoardo — APPROVED** [review]
LGTM, thanks @bundfussr 😄

File: tests/testthat/test-transform_range.R
Comment (bms63) [inline tests/testthat/test-transform_range.R:1]: Do we need a test for the default `outside_range`?  or is this covered?

File: R/transform_range.R [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+#'   *Permitted Values*: `TRUE`, `FALSE`
+#'
+#' @param outside_range Handling of values outside the source range
+#'
+#'   Values outside the source range (`source_range`) are transformed to `NA`.
+#'
+#'   If `"warning"` or `"error"` is specified, a warning or error is issued if
+#'   `source` includes any values outside the source range.
+#'
+#'   *Permitted Values*: `"NA"`, `"warning"`, `"error"`
```
Comment (bms63) [inline R/transform_range.R:34]: Is this set to "NA" or `NA_character` ?

File: R/compute_scale.R [hunk 6]
```diff
@@ -97,19 +97,16 @@ compute_scale <- function(source,
...
-      target <- (target + scale_constant) * scale_coefficient
-
-      if (flip_direction == TRUE) {
-        target <- max(target_range) - target
-      }
+      target <- transform_range(
+        target,
+        source_range = source_range,
+        target_range = target_range,
+        flip_direction = flip_direction
```
Comment (bms63) [inline R/compute_scale.R:108]: Should compute_scale inherit transform_range's documentation?

File: .lintr [hunk 7]
```diff
@@ -3,7 +3,7 @@ linters: linters_with_defaults(
     object_usage_linter=NULL,
     cyclocomp_linter(complexity_limit = 22),
     indentation_linter=NULL,
-    undesirable_function_linter = undesirable_function_linter()
+    undesirable_function_linter = undesirable_function_linter(symbol_is_undesirable = FALSE)
```
Comment (bms63) [inline .lintr:6]: ohhhh way to get rid of the pesky #nolints!!

**Review by bms63 — COMMENTED** [review]
Just a few thoughts!!

File: tests/testthat/test-transform_range.R
Comment (bundfussr) [reply tests/testthat/test-transform_range.R:1]: Test 1 and 2 use the default.

File: R/transform_range.R (hunk 5 above)
Comment (bundfussr) [reply R/transform_range.R:34]: The `outside_range` argument expects `"NA"`.

File: R/compute_scale.R (hunk 6 above)
Comment (bundfussr) [reply R/compute_scale.R:108]: The descriptions of the arguments differ. Thus they can't be inherited.

File: R/transform_range.R (hunk 5 above)
Comment (bms63) [reply R/transform_range.R:34]: SOrry!!!  So the argument `outside_range = "NA"` will produce NA values?

File: R/compute_scale.R (hunk 6 above)
Comment (bms63) [reply R/compute_scale.R:108]: but isn't this used in `compute_scale()` or am I missing something?

File: R/transform_range.R (hunk 5 above)
Comment (bundfussr) [reply R/transform_range.R:34]: Yes, values outside the source range result in `NA`. If `"warning"` or `"error"` is specified, a warning or error is issued in addition.

**Review by bms63 — APPROVED** [review]

**Comment by bms63** [comment]
Thanks @bundfussr !!

**Pull request merged**